File: packages/contracts/src/domain/weather.ts
// Weather contracts shared between the desktop main process (Open-Meteo proxy)
// and the renderer. Ported from lunofi-studio's weather module — keyless
// Open-Meteo, collapsed to an 8-bucket condition enum.

/**
 * Coarse weather condition. Open-Meteo's ~28 WMO interpretation codes collapse
 * to these buckets so the renderer only switches on a small set when picking a
 * glyph/icon.
 */
export type WeatherCondition =
  | 'clear'
  | 'partly_cloudy'
  | 'cloudy'
  | 'rain'
  | 'snow'
  | 'thunderstorm'
  | 'fog'
  | 'unknown';

export const WEATHER_CONDITIONS: readonly WeatherCondition[] = [
  'clear',
  'partly_cloudy',
  'cloudy',
  'rain',
  'snow',
  'thunderstorm',
  'fog',
  'unknown',
] as const;

export interface WeatherCurrent {
  /** Temperature in degrees Celsius (rounded to 0.1°). */
  tempC: number;
  condition: WeatherCondition;
  /** Short human label from the WMO table ("Clear sky", "Light rain", …). English. */
  label: string;
}

export interface GeocodeResult {
  lat: number;
  lon: number;
  /** Human label: "City, Country". */
  label: string;
}

/** Stable error code thrown when the upstream weather lookup fails. */
export const WEATHER_UNAVAILABLE = 'WEATHER_UNAVAILABLE';

File: crates/shiranami-core/src/models/weather.rs
//! Weather contracts, ported from `packages/contracts/src/domain/weather.ts`.
//!
//! Keyless Open-Meteo, with its ~28 WMO interpretation codes collapsed to the
//! eight buckets in [`WeatherCondition`] so the renderer switches over a small
//! set when picking a glyph.

use serde::{Deserialize, Serialize};
use specta::Type;
use specta_typescript::Number;

/// Coarse weather condition.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum WeatherCondition {
    /// Clear sky.
    Clear,
    /// Partly cloudy.
    PartlyCloudy,
    /// Overcast.
    Cloudy,
    /// Rain of any intensity.
    Rain,
    /// Snow of any intensity.
    Snow,
    /// Thunderstorm.
    Thunderstorm,
    /// Fog or depositing rime fog.
    Fog,
    /// An interpretation code outside the mapped set.
    Unknown,
}

impl WeatherCondition {
    /// Every condition, in the order the TypeScript `WEATHER_CONDITIONS` tuple
    /// declares them. Exposed so callers can enumerate the buckets without
    /// re-listing them and drifting.
    pub const ALL: [Self; 8] = [
        Self::Clear,
        Self::PartlyCloudy,
        Self::Cloudy,
        Self::Rain,
        Self::Snow,
        Self::Thunderstorm,
        Self::Fog,
        Self::Unknown,
    ];
}

/// Current conditions at a location.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct WeatherCurrent {
    /// Temperature in degrees Celsius, rounded to 0.1°.
    #[specta(type = Number)]
    pub temp_c: f64,
    /// Coarse condition bucket.
    pub condition: WeatherCondition,
    /// Short English label from the WMO table ("Clear sky", "Light rain", …).
    pub label: String,
}

/// A resolved place, as returned by the geocoding lookup.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct GeocodeResult {
    /// Latitude in decimal degrees.
    #[specta(type = Number)]
    pub lat: f64,
    /// Longitude in decimal degrees.
    #[specta(type = Number)]
    pub lon: f64,
    /// Human label, formatted "City, Country".
    pub label: String,
}
